redis/client.go

```diff
@@ -1077,15 +1079,41 @@ func (c *Client) connect() error {
 		return nil
 	}
 
-	// If k6 has a TLSConfig set in its state, use
-	// it has redis' client TLSConfig too.
-	if vuState.TLSConfig != nil {
-		c.redisOptions.TLSConfig = vuState.TLSConfig
-	}
+	tlsCfg := c.redisOptions.TLSConfig
+	if tlsCfg != nil {
+		if vuState.TLSConfig != nil {
+			// Merge k6 TLS configuration with the one we received from the
+			// Client constructor. This will need adjusting depending on which
+			// options we want to expose in the Redis module, and how we want
+			// the override to work.
+			tlsCfg.InsecureSkipVerify = vuState.TLSConfig.InsecureSkipVerify
+			tlsCfg.CipherSuites = vuState.TLSConfig.CipherSuites
+			tlsCfg.MinVersion = vuState.TLSConfig.MinVersion
+			tlsCfg.MaxVersion = vuState.TLSConfig.MaxVersion
+			tlsCfg.Renegotiation = vuState.TLSConfig.Renegotiation
+			tlsCfg.KeyLogWriter = vuState.TLSConfig.KeyLogWriter
+
+			tlsCfg.Certificates = append(tlsCfg.Certificates, vuState.TLSConfig.Certificates...)
+
+			// TODO: Merge vuState.TLSConfig.RootCAs with
+			// c.redisOptions.TLSConfig. k6 currently doesn't allow setting
+			// this, so it doesn't matter right now, but these should be merged.
+			// I couldn't find a way to do this with the x509.CertPool API
+			// though...
+		}
 
-	// use k6's lib.DialerContexter function has redis'
-	// client Dialer
-	c.redisOptions.Dialer = vuState.Dialer.DialContext
+		k6dialer, ok := vuState.Dialer.(*netext.Dialer)
+		if !ok {
+			panic(fmt.Sprintf("expected *netext.Dialer, got: %T", vuState.Dialer))
+		}
+		tlsDialer := &tls.Dialer{
+			NetDialer: &k6dialer.Dialer,
+			Config:    tlsCfg,
+		}
```

Comment:
Unfortunately I do think this completely circumvent the current tings netext.Dialer does:
1. Doing the resolving of the hostnames - including over `options.hosts`
2. blacklisting both hostnames and ips - which likely will be a big problem in cloud
3. emtting bytes read/written - which in this case might be an okay thing.

Looking at tls.Dialer though - it seems not possible to put netext.Dialer there. So maybe we should wrap it the other way aroudn :thinking:

Reply:
Good catch indeed 👍🏻  I gave a stab at addressing this concern in https://github.com/grafana/xk6-redis/pull/17/commits/3676befa7224342918319f71bcb5eca38c773218, by defining a custom dialer function which relies on k6's VU (`netext.Dialer`) dialer, and manually upgrades the connection to TLS (respecting the provided configuration) when TLS certificates are present.

Testing it showed that blocked hostnames were effectively taken into account.

Let me know if that addresses your point @mstoykov 🙇🏻